ethpm/utils/backend.py

```diff
@@ -0,0 +1,73 @@
+from typing import Generator, Type
+
+from eth_utils import to_tuple
+
+from ethpm.backends.base import BaseURIBackend
+from ethpm.backends.ipfs import (
+    DummyIPFSBackend,
+    InfuraIPFSBackend,
+    IPFSGatewayBackend,
+    LocalIPFSBackend,
+    get_ipfs_backend_class,
+)
+from ethpm.backends.registry import RegistryURIBackend
+from ethpm.exceptions import CannotHandleURI
+
+URI_BACKENDS = [
+    IPFSGatewayBackend,
+    InfuraIPFSBackend,
+    DummyIPFSBackend,
+    LocalIPFSBackend,
+    RegistryURIBackend,
+]
+
+
+def resolve_uri_contents(uri: str, fingerprint: bool = None) -> bytes:
+    is_resolvable = get_resolvable_backends_for_uri(uri)
+    is_translatable = get_translatable_backends_for_uri(uri)
+    if is_resolvable:
+        for backend in is_resolvable:
+            try:
+                contents = backend().fetch_uri_contents(uri)
+            except CannotHandleURI:
+                continue
+            return contents
+
+    if is_translatable:
+        if fingerprint:
+            raise CannotHandleURI(
+                "Registry URIs must point to a resolvable content-addressed URI."
+            )
+        for backend in is_translatable:
+            package_id = backend().fetch_uri_contents(uri)
+            return resolve_uri_contents(package_id, True)
+
+    raise CannotHandleURI(
+        "URI: {0} cannot be resolved by any of the available backends.".format(uri)
+    )
+
+
+@to_tuple
+def get_translatable_backends_for_uri(
+    uri: str
+) -> Generator[Type[BaseURIBackend], None, None]:
```

Comment:
Try replacing `Type[BaseURIBackend]` with:

```python
BackendType = typing.TypeVar('BackendType', bound=Type[BaseURIBackend)
```

I'm not sure if it will work, but the `bound` keyword IIRC is for specifying the lower bound on base classes that can be used.

Reply:
Also, cc @cburgdorf who has been our local type ninja.

Reply:
Haven't been able to get this working, even though after reading up on bound, it seems like this should work. 
with `BackendType = typing.TypeVar('BackendType', bound=Type[BaseURIBackend])`
I haven't been able to get around this error ... 
`(actual type "Type[BaseURIBackend]", expected type "BackendType")`
seems like `BackendType` isn't being evaluated properly? @cburgdorf any thoughts on what I'm doing wrong?